assignments/game-prototype.js

```diff
@@ -0,0 +1,323 @@
+    "d100": function() {return Math.floor(Math.random() * 100 + 1)},
+  }
+  
+  // === Constructors ===
+  
+  // == GameObject ==
+class GameObject {
+  constructor(data) {
+    this.createdAt = data.createdAt;
+    this.name = data.name;
+    this.dimensions = data.dimensions;
+    this.exists = true;
+  }
+  destroy() {
+    this.exists = false;
+    // return `${this.name} was removed from the game.`;
+  }
+}
+
+
+  // == CharacterStats ==
+class CharacterStats extends GameObject {
+  constructor(data) {
+    super(data);
+    this.healthPoints = data.healthPoints;
+  }
+  takeDamage(dam) {
+    this.healthPoints -= dam;
+    if (this.healthPoints <= 0) {
+      this.healthPoints = 0;
+      this.destroy()
+    };
+    return `${this.name} took ${dam} damage.`
+  }
+  displayHealth() {
+    return `${this.name} currently has ${this.healthPoints} hp remaining.`
+  };
+}
+
+
+  // == Humanoid ==
+
+class Humanoid extends CharacterStats {
+  constructor(data) {
+    super(data);
+    this.job = data.job;
+    this.weapons = data.weapons;
+    this.weaponDamage = data.weaponDamage;
+    this.stats = data.stats;
+    this.language = data.language;
+    this.preferredAtk = data.preferredAtk ? data.preferredAtk : this.basicAtk;
+  }
+
+  greet() {
+    return `${this.name} offers a greeting in ${this.language}.`;
+  };
+
+  attackStat() {
+    switch (this.weapons[0].weaponType) {
+      case "melee": return this.stats.str;
+      case "ranged": return this.stats.dex;
+      case "arcane": return this.stats.int;
+      case "divine": return this.stats.wis;
+      case "musical": return this.stats.cha;
+      default: return this.stats.str;
+    };
+  };
+
+  dodgeStat(dodger) {
+    switch (this.weapons[0].weaponType) {
+      case "melee": return dodger.stats.dex;
+      case "ranged": return dodger.stats.dex;
+      case "arcane": return dodger.stats.wis;
+      case "divine": return dodger.stats.wis;
+      case "musical": return dodger.stats.wis;
+      default: return dodger.stats.dex;
+    };
+  }
+
+  basicAtk(target) {
+    let atkAdj = "";
+    switch (this.job) {
+      case "hero":
+        atkAdj = " valiantly";
+        break;
+      case "villain":
+        atkAdj = " maliciously";
+        break;
+    };
+    if (target.exists){
+      if (this.exists) {
+        const dam = Math.ceil((this.attackStat() / 10) * this.weapons[0].weaponDamage * (Math.random() * .3 + .85));
+        const accuracy = (this.attackStat() / 25) * (1 - (this.dodgeStat(target) / 50));
+        const hit = accuracy >= Math.random();
+        return (hit) ? `${this.name} attacks ${target.name}${atkAdj} with ${this.weapons[0].weaponName}! ${target.takeDamage(dam)}` : `${this.name} attacks ${target.name}${atkAdj} with ${this.weapons[0].weaponName}! And misses!`;
+      } else {
+        return `${this.name} is dead!`
+      };
+    } else {
+      return `${target.name} has been vanquished by ${this.name}!`
+    };
+  };
```

Comment:
This is a bit of a mess to read, try to format and use comments to make this a bit easier to digest.